k8sClient.py

```diff
@@ -31,19 +38,170 @@ def get_nodes_with_label(self, label_selector: str) -> list[str]:
+        return [
+            "kubectl",
+            "--kubeconfig",
+            self._kc,
+            *namespace_args,
+            *self._oc_get_cmd(cmd),
+        ]
+
     def oc(
         self,
-        cmd: str,
+        cmd: str | Iterable[str],
         *,
         may_fail: bool = False,
         die_on_error: bool = False,
         namespace: typing.Optional[str] = None,
     ) -> host.Result:
-        namespace_args: tuple[str, ...] = ()
-        if namespace:
-            namespace_args = ("-n", namespace)
         return host.local.run(
-            ["kubectl", "--kubeconfig", self._kc, *namespace_args, *shlex.split(cmd)],
+            self._oc_get_full_cmd(cmd=cmd, namespace=namespace),
             die_on_error=die_on_error,
             log_level_fail=logging.DEBUG if may_fail else logging.ERROR,
         )
+
+    def oc_exec(
+        self,
+        cmd: str | Iterable[str],
+        *,
+        pod_name: str,
+        may_fail: bool = False,
+        die_on_error: bool = False,
+        namespace: typing.Optional[str] = None,
+    ) -> host.Result:
+        return self.oc(
+            ["exec", pod_name, "--", *self._oc_get_cmd(cmd)],
+            may_fail=may_fail,
+            die_on_error=die_on_error,
+            namespace=namespace,
+        )
+
+    def oc_get(
+        self,
+        what: str,
+        *,
+        may_fail: bool = False,
+        die_on_error: bool = False,
+        namespace: typing.Optional[str] = None,
+    ) -> typing.Optional[dict[str, typing.Any]]:
+        cmd = ["get", what, "-o", "json"]
+        ret = self.oc(
+            cmd,
+            may_fail=may_fail,
+            die_on_error=die_on_error,
+            namespace=namespace,
+        )
+
+        if not ret.success:
+            # No need for extra logging in this failure case. self.oc() already
+            # did all the logging we want.
+            return None
+
+        try:
+            data = json.loads(ret.out)
+        except ValueError:
+            data = None
+
+        if not isinstance(data, dict):
+            cmd_s = shlex.join(self._oc_get_full_cmd(cmd=cmd, namespace=namespace))
+            if not may_fail or die_on_error:
+                logger.error(
+                    f"Command {cmd_s} did not return a JSON dictionary but {ret.debug_str()}"
+                )
+            else:
+                logger.debug(f"Command {cmd_s} did not return a JSON dictionary")
+            if die_on_error:
+                sys.exit(-1)
+            return None
+
+        return data
+
+    def oc_debug(
+        self,
+        cmd: str | Iterable[str],
+        *,
+        node_name: str,
+        test_image: str,
+        namespace: str,
+        may_fail: bool = False,
+        die_on_error: bool = False,
+    ) -> host.Result:
+        container_name = f"ocp-tft-debug-{node_name}-{random.randint(0, 2**64-1):016x}"
+        # We use `kubectl debug` and not `oc debug`. There are thus some differences.
```

Comment:
I don't understand why we can't use the tools pod to do this by chrooting on to the host? 

Could this possibly work https://www.redhat.com/sysadmin/how-oc-debug-works ?

Reply:
But if there is an advantage of using kubectl debug, we can keep it as is.

But I am wondering if we can elevate the tools pod container to a level where it can run crictl properly.

Reply:
What I initially wanted to do, is get rid of crictl in `quay.io/wizhao/tft-tools:latest` image (for reasons mentioned the PR description).

I actually only later realized, that it would have been sufficient to change
```diff
diff --git i/pluginValidateOffload.py w/pluginValidateOffload.py
index c885cf34027d..1b8b427dd8a4 100644
--- i/pluginValidateOffload.py
+++ w/pluginValidateOffload.py
@@ -171,7 +171,7 @@ class TaskValidateOffload(PluginTask):
 
     def extract_vf_rep(self) -> Optional[str]:
         r = self.run_oc_exec(
-            f"crictl --runtime-endpoint=unix:///host/run/crio/crio.sock ps -a --name={self.perf_pod_name} -o json"
+            f"/host/usr/bin/crictl --runtime-endpoint=unix:///host/run/crio/crio.sock ps -a --name={self.perf_pod_name} -o json"
         )
 
         iface: Optional[str] = None
```
So yes, we can "use the tools pod to do this by chrooting on to the host" (or without CAP_SYS_CHROOT run the binary directly from `/host/usr/bin`).

---

But I didn't understand that at first, and wanted  to use `kubectl debug`. So I added the helpers like `K8sClient.oc_{get,exec,debug}()` and their `Task.run_oc_{get,exec,debug}()` counterparts.


> But I am wondering if we can elevate the tools pod container to a level where it can run crictl properly.

We could. `kubectl|oc debug` also just runs a pod. We could implement this in various ways, by running one pod or the other. `K8sClient.oc_debug()` does all the steps of creating and deleting the container in one and requires no additional setup. Granted, the only (current) user `PluginValidateOffset.extract_vf_rep()` already has a suitable pod at hand, so there is little to no benefit.

---

With all that I know now, I would not write `K8sClient.oc_debug()`. But as it is already written, I think it's a quite a useful function to have in the toolbox, as is has fewer requirements to the caller (that is, it doesn't require the caller to start/delete a suitable pod). What's also nice, is you see the `kubectl debug` command line in the log, which you can copy+paste to reproduce manually.

Reply:
ok. The problem is that this is just too slow. `oc delete` takes several seconds... (What the hell).

I close this PR, and:
- will use the host container we already for `kubectl exec ... chroot /host critctl ...`.
- reopen another PR with what I think are useful utility functions here.